src/NumericalAlgorithms/LinearOperators/MeanValue.cpp

```diff
@@ -10,29 +10,86 @@
 #include "Domain/Mesh.hpp"
 #include "ErrorHandling/Assert.hpp"
 
-template <>
-double mean_value_on_boundary(const DataVector& f, const Mesh<1>& mesh,
+double mean_value_on_boundary(const DataVector* /*unused*/,
+                              const DataVector& f, const Mesh<1>& mesh,
                               size_t d, Side side) noexcept {
   ASSERT(d == 0, "d = " << d);
   return Side::Lower == side ? f[0] : f[mesh.extents(0) - 1];
 }
 
 template <size_t Dim>
-double mean_value_on_boundary(const DataVector& f, const Mesh<Dim>& mesh,
+double mean_value_on_boundary(const gsl::not_null<DataVector*> boundary_buffer,
+                              const DataVector& f, const Mesh<Dim>& mesh,
                               size_t d, Side side) noexcept {
-  ASSERT(d < Dim, "d = " << d << ", Dim = " << Dim);
   const size_t N = mesh.extents(d);
   const Mesh<Dim - 1> mesh_on_boundary = mesh.slice_away(d);
-  DataVector f_on_boundary(mesh_on_boundary.number_of_grid_points());
   for (SliceIterator si(mesh.extents(), d, (Side::Lower == side ? 0 : N - 1));
        si; ++si) {
-    f_on_boundary[si.slice_offset()] = f[si.volume_offset()];
+    (*boundary_buffer)[si.slice_offset()] = f[si.volume_offset()];
   }
-  return definite_integral(f_on_boundary, mesh_on_boundary) /
+  return definite_integral(*boundary_buffer, mesh_on_boundary) /
          two_to_the(Dim - 1);
 }
 
+double mean_value_on_boundary(
+    const DataVector* /*unused*/,
+    const gsl::span<std::pair<size_t, size_t>> /*volume_and_slice_indices*/,
+    const DataVector& f, const Mesh<1>& mesh, const size_t d,
+    const Side side) noexcept {
+  ASSERT(d == 0, "d = " << d);
+  return Side::Lower == side ? f[0] : f[mesh.extents(0) - 1];
+}
+
+template <size_t Dim>
+double mean_value_on_boundary(
+    const gsl::not_null<DataVector*> boundary_buffer,
+    const gsl::span<std::pair<size_t, size_t>> volume_and_slice_indices,
+    const DataVector& f, const Mesh<Dim>& mesh, const size_t d,
+    const Side /*side*/) noexcept {
+  const Mesh<Dim - 1> mesh_on_boundary = mesh.slice_away(d);
+  for (const auto& volume_and_slice_index : volume_and_slice_indices) {
+    (*boundary_buffer)[volume_and_slice_index.second] =
+        f[volume_and_slice_index.first];
+  }
+  return definite_integral(*boundary_buffer, mesh_on_boundary) /
+         two_to_the(Dim - 1);
+}
+
+template <>
+double mean_value_on_boundary(const DataVector& f, const Mesh<1>& mesh,
+                              size_t d, Side side) noexcept {
+  ASSERT(d == 0, "d = " << d);
+  return Side::Lower == side ? f[0] : f[mesh.extents(0) - 1];
+}
+
+template <size_t Dim>
+double mean_value_on_boundary(const DataVector& f, const Mesh<Dim>& mesh,
+                              size_t d, Side side) noexcept {
+  ASSERT(d < Dim, "d = " << d << ", Dim = " << Dim);
```

Comment:
Would this `ASSERT` be better placed in the `not_null` version, so it is present for all execution paths?